For #9932 - Sticky header / footer in the menu

Mugurell opened the pull request:
Add support for having a specific menu item as a sticky header - won't be 
scrolled off screen to the top or as a sticky footer - won't be scrolled
offscreen past the list's bottom.
    
To use this the item intended to be sticky must be configured with the
"isSticky" property returning true.

https://user-images.githubusercontent.com/11428869/114214477-00de8480-996d-11eb-9993-d3c157da37f4.mp4



### After merge
- **Milestone**: Make sure issues closed by this pull request are added to the [milestone](https://github.com/mozilla-mobile/android-components/milestones) of the version currently in development.
- **Breaking Changes**: If this is a breaking change, please push a draft PR on [Reference Browser](https://github.com/mozilla-mobile/reference-browser) to address the breaking issues.

eliserichards commented on components/browser/menu/src/main/java/mozilla/components/browser/menu/BrowserMenu.kt:
```diff
@@ -41,6 +43,7 @@ open class BrowserMenu internal constructor(
     internal var isShown = false
     @VisibleForTesting
     internal lateinit var menuPositioningData: MenuPositioningData
+    internal var backgroundColor: Int = Color.RED
```
Nit: I'm guessing this can be removed 😋

Mugurell replied:
Will make it a lateinit var.

eliserichards commented on components/browser/menu/src/main/java/mozilla/components/browser/menu/BrowserMenu.kt:
```diff
@@ -124,19 +128,29 @@ open class BrowserMenu internal constructor(
-            menuList?.let {
-                showMenuBottom(it)
+        } else {
+            // The menu is by default set as a bottom one. Reconfigure it as a top one.
```
Can you add a comment above to the `// if` that it is for bottom toolbar?

eliserichards commented on components/browser/menu/src/main/java/mozilla/components/browser/menu/BrowserMenu.kt:
```diff
@@ -124,19 +128,29 @@ open class BrowserMenu internal constructor(
+                view.context, StickyItemPlacement.TOP
+            )
+
+            // An expandable bottom menu should always be scrolled to the top (default in LayoutManager).
+            // Otherwise try showing the bottom of the menu when not enough space to fit it on the screen.
```
If this `else` is for the top toolbar then these comments should be updated

Mugurell replied:
I think that is a correct description.
By default the menu is scrolled to the top. And this is what we want for the bottom menu. While for the top menu can be configured to start scrolled to the bottom.
Open to suggestion about what should say.

Mugurell replied:
If something is unclear it's better to fix it while we're at it. Reworded that to
> // By default the menu is laid out from and scrolled from top.
> // For the top menu it may be desired to initially show the bottom most items.
Hope this is better.

eliserichards commented on components/browser/menu/src/main/java/mozilla/components/browser/menu/BrowserMenu.kt:
```diff
@@ -200,6 +214,14 @@ open class BrowserMenu internal constructor(
 
+    private fun setColors(menuLayout: View, colorState: MenuStyle?) {
+        val listParent: CardView = menuLayout.findViewById<CardView>(R.id.mozac_browser_menu_menuView)
+        backgroundColor = colorState?.backgroundColor?.let {
+            listParent.setCardBackgroundColor(it)
+            it.defaultColor
+        } ?: listParent.cardBackgroundColor.defaultColor
```
Why do we need this `.let`? Can we just do this: 
```suggestion
        backgroundColor = listParent.cardBackgroundColor.defaultColor
```

Mugurell replied:
The `let` block is where we are also setting the background for the menu to be the one from the MenuStyle.
It's basically this code https://github.com/mozilla-mobile/android-components/blob/765ebf9829eb3622dddef964179ef285c2a7fc16/components/browser/menu/src/main/java/mozilla/components/browser/menu/BrowserMenu.kt#L70-L72 refacored.

eliserichards commented on components/browser/menu/src/main/java/mozilla/components/browser/menu/BrowserMenuAdapter.kt:
```diff
@@ -42,6 +44,16 @@ internal class BrowserMenuAdapter(
+    override fun isStickyItem(position: Int) = visibleItems[position].isSticky
+
+    override fun setupStickyItem(stickyItem: View) {
+        stickyItem.setBackgroundColor(menu!!.backgroundColor)
```
Nit: try not to use `!!`

eliserichards commented on components/browser/menu/src/main/java/mozilla/components/browser/menu/item/AbstractParentBrowserMenuItem.kt:
```diff
@@ -16,11 +16,14 @@ import mozilla.components.browser.menu.BrowserMenuItem
     private val endOfMenuAlwaysVisible: Boolean,
-    override val isCollapsingMenuLimit: Boolean = false
+    override val isCollapsingMenuLimit: Boolean = false,
+    override val isSticky: Boolean = false
```
I am thinking maybe we want default `= true` here, since we don't usually want things to scroll offscreen

eliserichards replied:
Same for: 
- BrowserMenuCategory.kt
- BrowserMenuCheckbox.kt
- BrowserMenuCheckbox.kt
- BrowserMenuCompoundButton.kt
- BrowserMenuHighlightableSwitch.kt
- BrowserMenuHighlightableItem.kt

eliserichards replied:
.. in all of the `BrowserMenuItem`s 😅 I realized there's a bunch more lol

Mugurell replied:
If all menu items fit the screen the list is not scrollable by default.
If not all menu items fit the screen they will have to be scrolled off the screen.
Having a sticky item is the exception - a sticky footer will not be scrolled downwards offscreen but all the items that do not fit can be scrolled off to the bottom.
I think the video in the PR description - especially the menu shown in landscape showcases best where the StickyLinearLayoutManager is useful and it's basically the fix for https://github.com/mozilla-mobile/fenix/issues/18591#issuecomment-804841491.

eliserichards commented on components/browser/menu/src/main/java/mozilla/components/browser/menu/view/StickyFooterLinearLayoutManager.kt:
```diff
@@ -0,0 +1,69 @@
+        offset: Int,
+        actuallyScrollToPositionWithOffset: (Int, Int) -> Unit
+    ) {
+        // The following scenarios are handled:
```
👏

eliserichards commented on components/browser/menu/src/main/java/mozilla/components/browser/menu/view/StickyFooterLinearLayoutManager.kt:
```diff
@@ -0,0 +1,69 @@
+ */
+open class StickyFooterLinearLayoutManager<T> constructor(
+    context: Context,
+    reverseLayout: Boolean = false
```
We have a sticky footer **and** a sticky header layout manager. Do we also need `reverseLayout`? What is the use case for this?

Mugurell replied:
`reverseLayout` is a constructor parameter of Android's `LinearLayoutManager`, one that AC clients might expect to be able to change to control whether the list should be layout from top to bottom or from bottom to top - effectively change the visible item positions - that same menu item might appear visually at the top of the list or at the bottom.
The list can be configured with this property without also having an item set as sticky.
Then sticky header/footer just prevents a specific item from being scrolled off to the top / bottom of the list.

eliserichards commented on components/browser/menu/src/main/java/mozilla/components/browser/menu/view/StickyHeaderLinearLayoutManager.kt:
```diff
@@ -0,0 +1,68 @@
+ */
+open class StickyHeaderLinearLayoutManager<T> constructor(
+    context: Context,
+    reverseLayout: Boolean = false
```
Same question here

eliserichards commented on components/browser/menu/src/main/java/mozilla/components/browser/menu/view/StickyItemLayoutManager.kt:
```diff
@@ -0,0 +1,452 @@
+import androidx.recyclerview.widget.RecyclerView
+import kotlinx.parcelize.Parcelize
+
+// Inspired from
+// https://github.com/qiujayen/sticky-layoutmanager/blob/b1ddb086db5b04ff3c5357dabe1bff47a935dd37/
+// sticky-layoutmanager/src/main/java/com/jay/widget/StickyHeadersLinearLayoutManager.java
+/**
```
Is there a specific way we need to format this to give them credit? @Amejia481 might know

Mugurell replied:
Not sure what needs to be changed.

Mugurell replied:
Will change "Inspired from" to "Credit to".

Amejia481 replied:
Feel free to use  "Inspired from" if you just take some ideas and create your own implementation, in the example above we ported the implementation.

eliserichards reviewed (commented):
This looks fantastic so far! I will do some thorough manual testing tomorrow 👏

Amejia481 reviewed (commented):
Looks good, great job 👍🏽 I will do some manual testing tomorrow so far so good!

Mugurell commented on components/browser/menu/src/main/java/mozilla/components/browser/menu/view/StickyHeaderLinearLayoutManager.kt:
```diff
@@ -0,0 +1,68 @@
+        // - the first shown item is translated upwards off screen
+        // (happens when [scrollToPositionWithOffset] was called with a big enough offset)
+        return getAdapterPositionForItemIndex(0) >= stickyItemPosition ||
+            getChildAt(0)?.bottom ?: 1 <= 0
```
If `getChildAt(0)` is null that will help return false.

Amejia481 replied:
Thanks for the clarification we can keep it is.

Mugurell replied:
Added a comment since in a month time we might forget about this reasoning.

Mugurell commented:
Thank you both for your help with this!
I've updated the patches based on your feedback and added tests.
I see this in a final form and ready to land.

Mugurell commented:
Seeing a
> This class implements Parcelable but does not provide a CREATOR field [ParcelCreator]
exception in `toolchain-linux64-android-gradle-dependencies`

Refactoring `internal data class SavedState` to `internal abstract class SavedState`

Mugurell commented:
Seeing an
> This class implements Parcelable but does not provide a CREATOR field [ParcelCreator]

exception in `toolchain-linux64-android-gradle-dependencies`
following the usage of `kotlinx.parcelize.Parcelize`
though the app functions as expected and in the decompiled app I'm seeing the expected methods

![image](https://user-images.githubusercontent.com/11428869/114702858-44e6d600-9d2d-11eb-925b-dde2c9bc8c41.png)

Mugurell commented:
Although the kotlin version should be enough to use the new `kotlinx.parcelize.Parcelize` seems there still some quircks to investigate.
Switched to using the now deprecated @Parcelize from `kotlin-android-extensions` like in the rest of the app for now.

Amejia481 reviewed (approved):
I tested locally and it's working like a charm!
Awesome work, this is a really sticky PR 😄
